File: videocode/input/media/Video.py
#!/usr/bin/env python3

from __future__ import annotations

import subprocess

from videocode.input.shape.Polygon import *
from videocode.constants import WORLD_TO_SCREEN_RATIO
from videocode.input.media.TrackedPath import TrackedPath


__all__ = ["Video", "TrackedPath"]


def _pixelToWorld(
    px: number, py: number,
    srcW: number, srcH: number,
    width: number, height: number,
    centerX: number, centerY: number,
) -> point:
    """
    Pure coordinate-conversion math backing `Video.track()` — factored out
    so it's testable without decoding any actual video. See `Video.track()`'s
    docstring for the full derivation; in short:

    `(px, py) / (srcW, srcH)` is a 0..1 fraction across the source frame.
    `position()` places the shape's bbox CENTER at `(centerX, centerY)` under
    the default `align=(0.5, 0.5)`, and world Y is positive-upward while
    image Y is positive-downward (X is not flipped), so:

        worldX = centerX + width  * (px / srcW - 0.5)
        worldY = centerY + height * (0.5 - py / srcH)
    """
    worldX = centerX + width * (px / srcW - 0.5)
    worldY = centerY + height * (0.5 - py / srcH)
    return (worldX, worldY)


class Video(Polygon):
    cppName = "Video"
    cppAttrs = Polygon.cppAttrs | {"filepath", "cuts", "speedRamps", "uvMapping", "uvAngle"}

    def __init__(
        self,
        filepath: str,
        cuts: list[frame | tuple[frame, frame]] = [],
        startFrame: frame = 0,
        endFrame: maybe[frame] = None,
        speedRamps: list[tuple[frame, frame, float]] = [],
        width: maybe[wunumber] = None,
        height: maybe[wunumber] = None,
        cornerRadius: percent = 0,
        strokeColor: rgba = TRANSPARENT,
        strokeWidth: wufloat = 0,
        uvMapping: UVMapping = UVMapping.STRETCH,
        uvAngle: wufloat = 0,
    ) -> None:
        """
        `cuts` are ranges of source-video frames to skip during playback —
        either a single frame index `n` (shorthand for `(n, n + 1)`) or a
        `(start, end)` pair cutting `[start, end)`.

        `startFrame`/`endFrame` restrict playback to the source frame range
        `[startFrame, endFrame)` — shorthand for cutting everything outside
        that range (`endFrame=None` plays to the end of the source).

        `speedRamps` are `(playbackStart, playbackEnd, rate)` triples that
        retime playback over a `[playbackStart, playbackEnd)` window —
        **playback-frame space**, i.e. the same post-cut index space
        `startFrame`/`endFrame` and the rest of the cut-down timeline already
        use, not raw source-file frame numbers. Within that window the source
        index advances at `rate` source-frames per playback-frame instead of
        the implicit 1x:

        - `1.0` (or no ramp covering a given frame): unchanged playback.
        - `> 1.0` (e.g. `2.0`): sped up — the source advances faster than
          playback, so more source footage passes in the same playback span.
        - `0.0 < rate < 1.0` (e.g. `0.5`): slow motion.
        - `0.0`: freeze-frame — the source index holds at the frame it was
          at when the window started.
        - negative (e.g. `-1.0`): reverse playback — the source index
          decreases as playback advances.

        Frame sampling is nearest-frame (no frame-blending) at every rate —
        matching the fact that there's no blend/lerp step anywhere else in
        `Video`'s frame pipeline; smooth interpolation for non-1x rates is a
        possible future enhancement, not implemented here.

        Ramp windows must be non-overlapping (in playback-frame space) —
        overlapping ramps raise a `ValueError`. A ramp window may cross a
        `cuts`/`startFrame`/`endFrame` boundary, but a cut boundary *inside*
        a non-1x ramp window isn't specially accounted for by the ramp's own
        rate math (only by where the ramp's source anchor sits) — ramps
        aren't expected to straddle a cut in practice.

        `uvMapping` controls how the texture is wrapped onto the shape —
        see the `UVMapping` enum for the mode semantics; `uvAngle` (degrees)
        rotates the angular origin of the polar modes.
        """
        self.filepath = filepath
        self.uvMapping = uvMapping
        self.uvAngle = uvAngle
        cuts = [c if isinstance(c, tuple) else (c, c + 1) for c in cuts]
        if startFrame:
            cuts.append((0, startFrame))
        if endFrame is not None:
            # the C++ side clamps `end` to the source's actual frame count,
            # so an over-large sentinel cuts everything past `endFrame`.
            cuts.append((endFrame, 2**31 - 1))
        self.cuts = cuts

        speedRamps = sorted(speedRamps, key=lambda ramp: ramp[0])
        for (aStart, aEnd, _), (bStart, bEnd, _) in zip(speedRamps, speedRamps[1:]):
            if bStart < aEnd:
                raise ValueError(
                    "Video speedRamps must be non-overlapping in playback space, "
                    f"got overlapping segments ({aStart}, {aEnd}) and ({bStart}, {bEnd})"
                )
        self.speedRamps = speedRamps

        # Rounding/stroking needs a known shape — if the caller didn't give
        # one, fall back to the video's natural frame size (ffprobe metadata,
        # no frame decode).
        if cornerRadius and width is None and height is None:
            out = subprocess.run(
                ["ffprobe", "-v", "error", "-select_streams", "v:0",
                 "-show_entries", "stream=width,height", "-of", "csv=p=0", filepath],
                capture_output=True, text=True, check=True,
            )
            w, h = out.stdout.strip().split(",")
            width = float(w) / WORLD_TO_SCREEN_RATIO
            height = float(h) / WORLD_TO_SCREEN_RATIO

        self.width = width
        self.height = height

        super().__init__(
            vertices=self.generateVertices(),
            fillColor=TRANSPARENT,
            strokeColor=strokeColor,
            strokeWidth=strokeWidth,
            cornerRadius=cornerRadius,
        )

    def generateVertices(self) -> list[point]:
        if self.width is None or self.height is None:
            return []
        return [(0, 0), (self.width, 0), (self.width, self.height), (0, self.height)]

    @prop(onSet=Polygon.updatePoints)
    def width() -> maybe[wunumber]: ...

    @prop(onSet=Polygon.updatePoints)
    def height() -> maybe[wunumber]: ...

    def track(
        self,
        x: number,
        y: number,
        startFrame: frame = 0,
        endFrame: maybe[frame] = None,
    ) -> TrackedPath:
        """
        Track a point through this video's raw source frames using OpenCV's
        Lucas-Kanade sparse optical flow (`cv2.calcOpticalFlowPyrLK`),
        starting from pixel `(x, y)` — the source video's NATIVE pixel
        resolution, plain image coordinates (top-left origin, y-down), NOT
        world units — at `startFrame`, forward through to `endFrame`
        (exclusive; `None` tracks to the last decodable frame of the source).

        Returns a `TrackedPath`: `{sourceFrameIndex: (worldX, worldY)}`,
        already converted into this `Video`'s world-space coordinates (see
        the conversion math below). Pass it to `someInput.attachTo(path)` to
        make any other `Input` follow the tracked point frame-by-frame —
        e.g. `text.attachTo(video.track(x, y))`.

        Coordinate conversion: `(px, py) / (srcW, srcH)` is a 0..1 fraction
        across the source frame. `position()` places this `Video`'s bbox
        CENTER at `self.meta.position` under the default `align=(0.5, 0.5)`
        (verified via `getTransformationMatrixFromMetadata` in
        `IVertexShader.hpp`: with a centered pivot the position/pivot terms
        cancel to a pure translation of the bbox center to `(x, y)`). World Y
        is positive-upward while image Y is positive-downward (row 0 = top,
        matching `parsePointsJson`'s `-y` flip and the "local space Y is
        flipped relative to world space" convention documented in
        `BezierPath.cpp`); X is not flipped. So:

            worldX = centerX + width  * (px / srcW - 0.5)
            worldY = centerY + height * (0.5 - py / srcH)

        where `(centerX, centerY)` is `self.meta.position` and `width`/
        `height` are this `Video`'s world-unit shape size (explicit
        `width=`/`height=` if given, else the same natural-size fallback
        `__init__` uses for `cornerRadius`: native pixel size divided by
        `WORLD_TO_SCREEN_RATIO`).

        Requires `opencv-python` (`pip install opencv-python`) — imported
        lazily here, mirroring this project's lazy-import convention, so
        scripts that never call `.track()` don't need it installed at all;
        the C++ backend already links OpenCV, but there is no Python-side
        `cv2` dependency anywhere else in `videocode`.

        Tracking quality: LK optical flow degrades on fast motion, occlusion,
        featureless regions, or large lighting changes. When OpenCV reports
        lost tracking (`status == 0` for a frame), this holds the last
        known-good position for that and every following frame rather than
        raising or extrapolating — a static-but-not-silently-wrong result
        beats a crash. Inspect the returned `TrackedPath` if you suspect a
        lost track.

        Known V1 limitations:

        - Frame indices are in **source-frame space** — the same raw index
          space `cuts`/`startFrame`/`endFrame` cut against, NOT the post-cut/
          post-speedRamp playback timeline. If this `Video` also uses `cuts`
          or `speedRamps`, a tracked source-frame index doesn't automatically
          correspond to the same playback frame; realigning tracked
          positions to a retimed playback timeline is a possible future
          enhancement, not implemented here (mirrors the caveat already
          documented on `speedRamps` above).
        - The conversion assumes this `Video` is stationary: `self.meta.
          position` is read once, at call time, and baked into every
          returned position. If the `Video` itself moves during playback
          (`.position()`/`.moveTo()`/etc. applied to it over time), tracked
          positions won't follow — this is a one-shot snapshot of wherever
          the `Video` sits when `.track()` runs.
        - Seeking to `startFrame` uses OpenCV's `CAP_PROP_POS_FRAMES`, which
          is only as accurate as the container's keyframe index for some
          codecs. Reads from `startFrame` onward are sequential (no further
          seeking), which is the reliable part of the pipeline.
        """
        try:
            import cv2
            import numpy as np
        except ImportError as e:
            raise ImportError(
                "Video.track() needs OpenCV on the Python side — install it "
                "with `pip install opencv-python` (this is intentionally not "
                "a hard dependency of videocode: only .track() callers need it)."
            ) from e

        cap = cv2.VideoCapture(self.filepath)
        if not cap.isOpened():
            raise RuntimeError(f"Video.track(): OpenCV could not open '{self.filepath}'")

        try:
            srcW = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
            srcH = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
            if srcW <= 0 or srcH <= 0:
                raise RuntimeError(f"Video.track(): OpenCV reported an invalid frame size for '{self.filepath}'")

            if endFrame is None:
                frameCount = int(cap.get(cv2.CAP_PROP_FRAME_COUNT))
                endFrame = frameCount if frameCount > startFrame else startFrame + 1

            width = self.width if self.width is not None else srcW / WORLD_TO_SCREEN_RATIO
            height = self.height if self.height is not None else srcH / WORLD_TO_SCREEN_RATIO

            # Snapshot — see the "assumes stationary" limitation above.
            centerX, centerY = self.meta.position.x, self.meta.position.y

            def toWorld(px: float, py: float) -> point:
                return _pixelToWorld(px, py, srcW, srcH, width, height, centerX, centerY)

            cap.set(cv2.CAP_PROP_POS_FRAMES, startFrame)
            ok, frameImg = cap.read()
            if not ok:
                raise RuntimeError(f"Video.track(): could not read startFrame={startFrame} from '{self.filepath}'")
            prevGray = cv2.cvtColor(frameImg, cv2.COLOR_BGR2GRAY)

            winSize = (21, 21)
            maxLevel = 3
            criteria = (cv2.TERM_CRITERIA_EPS | cv2.TERM_CRITERIA_COUNT, 30, 0.01)

            pts = np.array([[[float(x), float(y)]]], dtype=np.float32)
            positions: dict[frame, point] = {startFrame: toWorld(x, y)}

            frameIdx = startFrame
            while frameIdx + 1 < endFrame:
                ok, frameImg = cap.read()
                if not ok:
                    break
                gray = cv2.cvtColor(frameImg, cv2.COLOR_BGR2GRAY)
                nextPts, status, _err = cv2.calcOpticalFlowPyrLK(
                    prevGray, gray, pts, np.empty_like(pts),
                    winSize=winSize, maxLevel=maxLevel, criteria=criteria,
                )
                frameIdx += 1
                if nextPts is not None and status is not None and status[0][0] == 1:
                    pts = nextPts
                # else: hold the last known-good pts (lost track — see docstring).
                px, py = float(pts[0][0][0]), float(pts[0][0][1])
                positions[frameIdx] = toWorld(px, py)
                prevGray = gray

            return TrackedPath(positions)
        finally:
            cap.release()

File: src/input/media/Video.cpp
/*
** EPITECH PROJECT, 2025
** video-code
** File description:
** Video
*/

#include "input/media/Video.hpp"

#include <algorithm>
#include <cmath>
#include <opencv2/imgproc.hpp>

#include "utils/Exception.hpp"
#include "vulkan/MeshFactory.hpp"

Video::Video(json::object_t&& args)
    : BezierPath(std::move(args))
{
    std::string filepath = _baseArgs.at("filepath");

    _video.open(filepath);

    if (!_video.isOpened()) {
        throw Error("Could not load Video: " + filepath);
    }

    _nbFrame = static_cast<size_t>(_video.get(cv::CAP_PROP_FRAME_COUNT));

    if (_nbFrame == 0) {
        throw Error("Video has no frames: " + filepath);
    }

    // Parse, clamp and merge the requested cut ranges into sorted, non-overlapping
    // [start, end) pairs so mapToSourceIndex can walk them in a single forward pass.
    if (_baseArgs.contains("cuts")) {
        for (const auto& raw : _baseArgs.at("cuts")) {
            auto   pair = raw.get<std::vector<size_t>>();
            size_t start = std::min(pair[0], _nbFrame);
            size_t end = std::min(pair[1], _nbFrame);

            if (end > start) {
                _cuts.push_back({start, end});
            }
        }
    }

    std::sort(_cuts.begin(), _cuts.end());

    std::vector<std::pair<size_t, size_t>> merged;
    for (const auto& cut : _cuts) {
        if (!merged.empty() && cut.first <= merged.back().second) {
            merged.back().second = std::max(merged.back().second, cut.second);
        } else {
            merged.push_back(cut);
        }
    }
    _cuts = std::move(merged);

    size_t totalCut = 0;
    for (const auto& [start, end] : _cuts) {
        totalCut += end - start;
    }
    _playbackLength = (totalCut < _nbFrame) ? (_nbFrame - totalCut) : 1;

    // Speed ramps are expressed in playback space (the same post-cut index space
    // `_playbackLength` describes), so they can only be parsed/clamped once
    // `_playbackLength` is known. `end` is clamped to `_playbackLength` the same
    // way cuts clamp to `_nbFrame` above.
    if (_baseArgs.contains("speedRamps")) {
        for (const auto& raw : _baseArgs.at("speedRamps")) {
            auto   triple = raw.get<std::vector<double>>();
            size_t start = std::min(static_cast<size_t>(triple[0]), _playbackLength);
            size_t end = std::min(static_cast<size_t>(triple[1]), _playbackLength);
            double rate = triple[2];

            if (end > start) {
                _speedRamps.push_back({start, end, rate});
            }
        }
    }

    // The Python side already rejects overlapping segments; sorting here just
    // lets mapToSourceIndex do a simple ordered scan.
    std::sort(_speedRamps.begin(), _speedRamps.end(),
              [](const SpeedRamp& a, const SpeedRamp& b) { return a.playbackStart < b.playbackStart; });

    _lastIndex = mapToSourceIndex(0);
    _currentFrame = getFrameAt(_lastIndex);
}

// The original cuts-only mapping: shifts forward over every cut range that the
// (progressively shifted) index has reached. Cuts are sorted and non-overlapping,
// so a single forward pass suffices. Kept standalone (rather than folded into
// mapToSourceIndex) so: (a) cuts-only behavior — no speedRamps at all — stays
// byte-for-byte identical to before speed ramps existed, and (b) a speed-ramp
// segment can anchor its own rate math on "what the plain cuts-only mapping
// would have produced at the start of this ramp".
size_t Video::mapCutsOnly(size_t playbackIndex) const
{
    size_t source = playbackIndex;

    for (const auto& [start, end] : _cuts) {
        if (source >= start) {
            source += end - start;
        } else {
            break;
        }
    }

    return source;
}

// Maps a position in the cut-down playback timeline back to the corresponding
// source-video frame index. Piecewise: `_cuts` and `_speedRamps` both carve up
// playback space, but only `_speedRamps` need special per-segment rate math —
// everywhere else (including a `rate == 1.0` ramp, which is a documented no-op)
// the plain cuts-only additive mapping applies unchanged.
//
// A ramp segment anchors on `mapCutsOnly(playbackStart)` — the source frame the
// plain mapping would show at the start of the ramp window — then advances by
// `round((playbackIndex - playbackStart) * rate)` source frames from there:
// rate 1 preserves the anchor+delta identity, 0 freezes on the anchor frame,
// negative rates walk the source index backwards (reverse playback), and the
// result is clamped into the video's valid source-frame range at the edges.
//
// Known limitation: a cut boundary falling *inside* a non-1x ramp window isn't
// accounted for by the ramp's own math (only by the anchor, which is computed
// once at the ramp's start) — segments aren't expected to straddle a cut in
// practice, and this mirrors the "no frame-blending" simplicity of the rest of
// this feature.
size_t Video::mapToSourceIndex(size_t playbackIndex) const
{
    for (const auto& ramp : _speedRamps) {
        if (playbackIndex < ramp.playbackStart) {
            break; // sorted ascending — no later ramp can contain this index either
        }
        if (playbackIndex >= ramp.playbackEnd) {
            continue;
        }
        if (ramp.rate == 1.0) {
            break; // no-op ramp: fall through to the plain additive mapping below
        }

        size_t    anchor = mapCutsOnly(ramp.playbackStart);
        long long delta = static_cast<long long>(playbackIndex - ramp.playbackStart);
        long long offset = (ramp.rate == 0.0) ? 0 : static_cast<long long>(std::llround(static_cast<double>(delta) * ramp.rate));
        long long source = static_cast<long long>(anchor) + offset;

        if (source < 0) {
            source = 0;
        } else if (source >= static_cast<long long>(_nbFrame)) {
            source = static_cast<long long>(_nbFrame) - 1;
        }

        return static_cast<size_t>(source);
    }

    return mapCutsOnly(playbackIndex);
}

cv::Mat Video::getFrameAt(size_t index)
{
    if (index >= _nbFrame) {
        index = _nbFrame - 1;
    }

    cv::Mat frame;

    while (true) {
        _video.set(cv::CAP_PROP_POS_FRAMES, static_cast<double>(index));
        _video.read(frame);

        if (frame.empty()) {
            if (index == 0) {
                throw Error("Video has no readable frames: " + _baseArgs.at("filepath").get<std::string>());
            }
            index--;
        } else {
            break;
        }
    }

    if (frame.channels() != 4) {
        cv::cvtColor(frame, frame, cv::COLOR_BGR2BGRA);
    }

    return frame;
}

Mesh Video::getMesh(const Metadata& meta, const Config& config)
{
    size_t playbackIndex = meta.frameIndex;

    if (playbackIndex >= _playbackLength) {
        playbackIndex = _playbackLength - 1;
    }

    size_t index = mapToSourceIndex(playbackIndex);

    if (index != _lastIndex) {
        _currentFrame = getFrameAt(index);
        _lastIndex = index;
        if (_reupload) {
            _reupload(_currentFrame);
        }
    }

    bool hasCustomShape = meta.pointsPtr && meta.pointsPtr->size() >= 4 && meta.pointsPtr->size() % 2 == 0;
    if (hasCustomShape)
        return BezierPath::getMesh(meta, config);

    float w = static_cast<float>(_currentFrame.cols);
    float h = static_cast<float>(_currentFrame.rows);

    if (meta.args().contains("width")) w = meta.args().at("width").get<float>();
    if (meta.args().contains("height")) h = meta.args().at("height").get<float>();

    MeshFactory factory({w, h}, meta, config);

    float    opacity = meta.opacity / 255.f;
    uint32_t base = factory.vertexCount();
    factory.addVertex(0.f, 0.f, 0.f, 0.f, opacity);
    factory.addVertex(w, 0.f, 1.f, 0.f, opacity);
    factory.addVertex(w, h, 1.f, 1.f, opacity);
    factory.addVertex(0.f, h, 0.f, 1.f, opacity);

    factory.addIndex(base + 0);
    factory.addIndex(base + 1);
    factory.addIndex(base + 2);
    factory.addIndex(base + 0);
    factory.addIndex(base + 2);
    factory.addIndex(base + 3);

    Mesh mesh = factory.generateMesh();
    mesh.hasTexture = true;
    mesh.textureDescriptor = _descriptor;
    return mesh;
}
